File: HackerRank/The Maximum Subarray.cpp
/*
  The Maximum Subarray (50 Points)
  HackerRank
  Esteban Arango Medina

  Solution:
		This is the perfect use case for the Kadane's algorithm. The only tricky part is
		the sum of Non-contiguous (not necessarily contiguous) subarray when all of them
		are negatives, in that case you just print the max sum of the contiguous subarray twice.

  Notes.
    The Kadane's algorithm
    	http://codeforces.com/blog/entry/13713
*/
#include <cmath>
#include <cstdio>
#include <vector>
#include <iostream>
#include <algorithm>
using namespace std;
typedef vector <int> vi;
#define MAX 10000

int main() {
	// freopen("in.in", "r", stdin);

  int cases;scanf("%d",&cases);
  while(cases--) {
  	int n, current_max, max_contiguous_sum, max_sum;
  	max_sum = 0;
  	current_max = max_contiguous_sum = (-1 * MAX);
  	scanf("%d",&n);
  	for (int a, i = 0; i < n; ++i) {
  		scanf("%d",&a);
  		current_max = max(a,current_max + a);
  		max_contiguous_sum = max(current_max,max_contiguous_sum);
  		if (a>0){
  			max_sum += a;
  		}
  	}
  	if (max_contiguous_sum < 0)
  		max_sum = max_contiguous_sum;
  	printf("%d %d\n",max_contiguous_sum, max_sum);
  }
  return 0;
}

File: HackerRank/The Maximum Subarray.rb

file = File.exist?('in.in') ? File.new('in.in', 'r') : STDIN
MAX = 10000

cases = file.gets.chomp.to_i
cases.times do
	file.gets.chomp # no need this
	numbers = file.gets.chomp
  max_sum = 0;
  current_max = max_contiguous_sum = (-1 * MAX);
	numbers.split(' ').each do |a|
		a = a.to_i
		current_max = [a,current_max + a].max
		max_contiguous_sum = [current_max,max_contiguous_sum].max
		max_sum += a if a > 0
	end
 	max_sum = max_contiguous_sum if (max_contiguous_sum < 0)
 	puts "#{max_contiguous_sum} #{max_sum}"
end